cli/azd/pkg/pipeline/github_provider.go

```diff
@@ -627,18 +619,85 @@ func (p *GitHubCiProvider) configurePipeline(
 	provisioningProvider provisioning.Options,
 	additionalSecrets map[string]string,
 	additionalVariables map[string]string,
+	options *ConfigOptions,
 ) (CiPipeline, error) {
-
 	repoSlug := repoDetails.owner + "/" + repoDetails.repoName
+
+	// Variables and Secrets for a gh-actions are independent from the gh-action. They are set on the repository level.
+	// We need to clean up the previous values before setting the new ones.
+	// By doing this, we are handling:
+	// - When a secret is moved to be a variable (or vice versa). Don't leak the previous value on the pipeline.
+	// - When there was a previous additional variable/secret set and then it was updated to empty string or unset from .env.
+	msg := ""
+	var procErr error
+	ciSecrets, ciVariables := []string{}, []string{}
+	if len(options.Variables) > 0 || len(options.Secrets) > 0 {
+		msg = "Setting up project's secrets and variables to be used in the pipeline"
+		ciSecretsInstance, err := p.ghCli.ListSecrets(ctx, repoSlug)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get list of repository secrets: %w", err)
+		}
+		ciVariablesInstance, err := p.ghCli.ListVariables(ctx, repoSlug)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get list of repository variables: %w", err)
+		}
+		ciSecrets = ciSecretsInstance
+		ciVariables = ciVariablesInstance
+		p.console.ShowSpinner(ctx, msg, input.Step)
+	}
+
+	defer func() {
+		if msg != "" {
+			p.console.StopSpinner(ctx, msg, input.GetStepResultFormat(procErr))
+		}
+		if procErr == nil {
+			p.console.MessageUxItem(ctx, &ux.MultilineMessage{
+				Lines: []string{
+					"",
+					"GitHub Action secrets are now configured. You can view GitHub action secrets that were created at this link:",
+					output.WithLinkFormat("https://github.com/%s/settings/secrets/actions", repoSlug),
+					""},
+			})
+		}
+	}()
+
+	// iterate the existing secrets on the pipeline and remove the ones matching the project's secrets or variables
+	for _, existingSecret := range ciSecrets {
+		for _, key := range options.Variables {
+			if existingSecret == key {
+				_ = p.ghCli.DeleteSecret(ctx, repoSlug, key)
+			}
+		}
+		for _, key := range options.Secrets {
+			if existingSecret == key {
+				_ = p.ghCli.DeleteSecret(ctx, repoSlug, key)
+			}
+		}
+	}
+	// iterate the existing variables on the pipeline and remove the ones matching the project's secrets or variables
+	for _, existingVariable := range ciVariables {
+		for _, key := range options.Variables {
+			if existingVariable == key {
+				_ = p.ghCli.DeleteVariable(ctx, repoSlug, key)
+			}
+		}
+		for _, key := range options.Secrets {
+			if existingVariable == key {
+				_ = p.ghCli.DeleteVariable(ctx, repoSlug, key)
+			}
+		}
+	}
```

Comment:
Why do we need/want to delete matching variables/secrets vs just setting them like we previously were? Surely Setting a variable/secret will overwrite the existing value?

Reply:
> just setting them like we previously were?

We still do this.  The problem is what happen on a flow like this:
- User defines FOO in azure.yaml as a secret
- user calls `azd env set FOO bar` and then `azd pipeline config`
- azd sets FOO=bar as a secret on the repository
- later, the user runs `azd env unset FOO`, removing FOO from the .env and runs `azd pipeline config` expecting to update the .env after this change.
- Since FOO is not in .env anymore, azd doesn't know that it was or not set on a previous run... THAT's why is azd finds FOO as a secret or variable in the repo, it is deleted. Ensuring the customer's expectation.

The other scenario is when, instead of calling `azd env unset FOO`, customer just calls `azd env set FOO ""`.
The empty string is persisted to the .env, but GitHub does not allow secrets/variables with empty string.  So, for github provider, env vars from .env with empty string value are considered by azd as not set.
When a ENV VAR is referenced in the gh-action, if it is not set, it will return an empty string

Reply:
I don't think we can say it is the expectation of the customer to delete secrets that are not defined in azd.  What if there are additional github env vars set outside of azd?  We can't know this.

Reply:
> What if there are additional github env vars set outside of azd? We can't know this.

azd will never delete a secret or variable from a repository if the name of such secret/variable is not listed in `azure.yaml`.

Customer explicitly set the list of variables/secrets on `azure.yaml` to delegate azd to sync those to their CI/CD when running `azd pipeline config`